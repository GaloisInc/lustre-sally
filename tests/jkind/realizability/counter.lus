-- Unrealizable

node main(x, y, z : int) returns (ok : bool);
var
  counter : int;
let
  counter = 0 -> 1 + pre counter;
  assert counter < 5 => x < y - 1;
  
  ok = x < z and z < y;
  --%PROPERTY ok;
  --%REALIZABLE x, y;
tel;
