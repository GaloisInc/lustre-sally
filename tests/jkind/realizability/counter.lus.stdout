Core Lustre
===========

node (x : int, y : int, z : int)
  returns (ok)
  assumes (Implies)
  shows (ok)
let
  ok : bool
    = And(Lt(x, z), Lt(z, y))
   
  pre_counter : int
    = pre counter
   
  counter : int
    = 0 -> Add(1, pre_counter)
   
  Implies : bool
    = Implies(Lt(counter, 5), Lt(x, Sub(y, 1)))
tel
Sally Model
===========

(define-state-type
   S
   (
     (Implies Bool)
     (counter Int)
     (ok Bool)
     (x Int)
     (y Int)
     (z Int)
     (|counter:init| Bool)
     (|gal-initializing| Bool)
     (|pre_counter| Int))
   (
     (x Int)
     (y Int)
     (z Int)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |counter:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.x input.x)
      (= next.y input.y)
      (= next.z input.z)
      next.Implies
      (=
         next.ok
         (and
            (< next.x next.z)
            (< next.z next.y)))
      (= next.|pre_counter| state.counter)
      (ite
         state.|counter:init|
         (= next.counter 0)
         (=
            next.counter
            (+ 1 next.|pre_counter|)))
      (= next.|counter:init| false)
      (=
         next.Implies
         (=>
            (< next.counter 5)
            (<
               next.x
               (- next.y 1))))))

(query
   TS
   (or |gal-initializing| ok))

[Lustre] Validating properties:
[Lustre] Property ok... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
[Invalid] 
Trace
=====

ok:
 Step | x | y | z | |-> | ok    
------+---+---+---+-----+-------
 1    | 0 | 2 | 0 |     | False 

[Lustre] Model status: [Invalid] 
